File: go/src/chapter_4/01_bfs.go
// Для заданного направленного графа разработайте алгоритм, проверяющий
// существование маршрута между двумя узлами.
package chapter_4

// used 'breadth first search' algorithm
func bfs(graph map[int][]int, start, end int) bool {
	visited, queue := []int{}, GetQueue()
	visited = append(visited, start)
	queue.Add(start)

	for {
		node, err := queue.Remove()
		if err != nil {
			break
		}
		for _, neighbour := range graph[node] {
			if ok := find(visited, neighbour); !ok {
				if neighbour == end {
					return true
				}
				visited = append(visited, neighbour)
				queue.Add(neighbour)
			}
		}
	}
	return false
}

func find(v []int, n int) bool {
	for _, i := range v {
		if i == n {
			return true
		}
	}
	return false
}

File: python/chapter_4/01_bfs.py
# Для заданного направленного графа разработайте алгоритм, проверяющий
# существование маршрута между двумя узлами.
import collections
import unittest


# used 'breadth first search' algorithm
def bfs(graph, start, end):
    visited, queue = set(), collections.deque([start])
    visited.add(start)
    while queue:
        node = queue.popleft()
        for neighbour in graph[node]:
            if neighbour not in visited:
                if neighbour == end:
                    return True
                visited.add(neighbour)
                queue.append(neighbour)
    return False


class Test(unittest.TestCase):
    graph = {0: [1, 2], 1: [2], 2: [3], 3: [1, 2]}

    def test_bfs_true(self):
        self.assertTrue(bfs(self.graph, 0, 2))

    def test_bfs_false(self):
        self.assertFalse(bfs(self.graph, 1, 4))


if __name__ == "__main__":
    unittest.main()
